tests/tests/code_macro_app.plf

```diff
@@ -0,0 +1,579 @@
+    (f_re.range (arrow String (arrow String RegLan)))
+    (f_re.++ (arrow RegLan (arrow RegLan RegLan)))
+    (f_str.to_re (arrow String RegLan))
+    (f_re.inter (arrow RegLan (arrow RegLan RegLan)))
+    (f_re.union (arrow RegLan (arrow RegLan RegLan)))
+    ((f_re.loop n1 n2) (arrow RegLan RegLan))
+    (f_str.in_re (arrow String (arrow RegLan Bool)))
+    
+    ; ...
+    ; fail: unknown term
+  )
+)
+(program well_typed ((t term)) Ok
+  (match t
+    ; special case: the equality symbol itself is well-typed
+    (f_= ok)
+    (default 
+      (do (type_check t) ok))))
+; Congruence :
+
+(declare refl (! t term
+              (! r (^ (well_typed t) ok)
+	(holds (= t t)))))
+
+(declare symm (! s term
+        (! t term
+        (! u (holds (= s t))
+        (holds (= t s))))))
+
+(declare neg_symm (! s term
+        (! t term
+        (! u (holds (not (= s t)))
+        (holds (not (= t s)))))))
+
+(declare trans (! t1 term
+        (! t2 term
+        (! t3 term
+        (! u1 (holds (= t1 t2))
+        (! u2 (holds (= t2 t3))
+        (holds (= t1 t3))))))))
+
+(declare cong (! a1 term
+              (! b1 term
+              (! a2 term
+              (! b2 term
+              (! u1 (holds (= a1 b1))
+              (! u2 (holds (= a2 b2))
+              (! t (^ (well_typed (apply a1 a2)) ok)
+                (holds (= (apply a1 a2) (apply b1 b2)))))))))))
+
+(declare true_intro (! f term
+                    (! u (holds f)
+                    (holds (= f true)))))
+
+(declare true_elim (! f term
+                   (! u (holds (= f true))
+                   (holds f))))
+
+(declare false_intro (! f term
+                    (! u (holds (not f))
+                    (holds (= f false)))))
+
+(declare false_elim (! f term
+                   (! u (holds (= f false))
+                   (holds (not f)))))
+
+
+; head and tail for n-ary operators with null terminators
+(program str.head ((t term)) term
+  (match t ((apply t1 t2) (match t1 ((apply f_str.++ t12) t12)))))
+(program str.tail ((t term)) term
+  (match t ((apply f_str.++ t2) t2)))
+
+;t = x, s=y ---> (str.++ x y)
+;t = (str.++ x1 x2), s = y1 ----> (str.++ x1 (str.++ x2 y1))
+;t = (str.++ (str.++ x1 x2) (str.++ y1 y2)), s=y ----> (str.++ (str.++ (str.++ (str.++ x1 x2) y1) y2) y)
+(program str.concat ((t term) (s term)) term
+  (match t 
+    ((apply t1 t2)
+      (match t1 ((apply f_str.++ t12) (apply t1 (str.concat t2 s)))))
+    (emptystr s))
+)
+
+
+(program fun ((t term) (s term)) term
+         (and t s)
+         )
+
+(declare and_pf
+              (! a term
+              (! b term
+              (! c term
+              (! a_pf (holds a)
+              (! b_pf (holds b)
+              (! sc (^ (fun a b) c)
+                 (holds c))))))))
+
+(check-assuming
+  ((: a term)
+   (: b term)
+   (: a_pf (holds a))
+   (: b_pf (holds b)))
+  (holds (and a b))
+  (and_pf _ _ _ a_pf b_pf))
+
+    
+    
+;(program sc_string_reduction ((t term) (k term)) term
```

Comment:
Let's just delete this.

Reply:
Fair enough. I'll do you one better. eb113b5